### PR TITLE
[Program Migration] Support export and import of program definitions

### PR DESCRIPTION
**Pull request by caitlinshk**
### Description

This PR does a few things to let us export and import program definitions:
 - Updates the `ProgramDefinition` class so that it can be serialized into JSON and deserialized from JSON back into a `ProgramDefinition` object.
 - Updates `AdminExportController` to take the selected program, serialize it into JSON, and download that JSON file.
 - Updates `AdminImportController` to read the uploaded JSON file, deserialize it into a `ProgramDefinition` object, and display the program's blocks and questions on screen.

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | enhancement | bug | under-development | dependencies | infrastructure | ignore-for-release | database >
- [x] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers)
- [x] Added an additional reviewer from outside your organization as FYI (if the primary reviewer is in the same organization as you)
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)

#### User visible changes

- [x] Manually tested at 200% size
- [x] Manually evaluated tab order

### Instructions for manual testing

1. Enable the PROGRAM_MIGRATION_ENABLED flag.
2. As an admin, open the Export tab.
3. Choose a program and click Download.
4. Open the Import tab.
5. Upload the file you just downloaded and click Upload.
6. Verify the page updates and shows the program you downloaded with its blocks and questions. Note that we only show the question IDs right now. A future PR will show all the question data.

### Issue(s) this completes

Part of epic #7087

### CODE REVIEW

**Comment by caitlinshk** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/civiform/civiform/7113?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#7113** <a href="https://app.graphite.dev/github/pr/civiform/civiform/7113?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#7096** <a href="https://app.graphite.dev/github/pr/civiform/civiform/7096?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7090** <a href="https://app.graphite.dev/github/pr/civiform/civiform/7090?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7094** <a href="https://app.graphite.dev/github/pr/civiform/civiform/7094?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @caitlinshk and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: server/app/controllers/admin/ProgramMigration.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  @JsonCreator
+  public ProgramMigration(@JsonProperty("program") ProgramDefinition program) {
+    this.program = program;
+  }
+
+  public ProgramDefinition getProgram() {
+    return program;
+  }
+
+  public void setPrograms(ProgramDefinition programs) {
```
Comment (rockycodes) [inline server/app/controllers/admin/ProgramMigration.java:28]: question (non-blocking): should this be plural?

**Review by rockycodes — COMMENTED** [review]
I didn't do an in-depth review since I'm the fyi reviewer, but this work is so cool!! I just left one tiny comment that stood out to me.

File: server/app/services/program/ProgramDefinition.java [hunk 2]
```diff
@@ -30,8 +33,22 @@
 import services.question.types.QuestionDefinition;
 import services.question.types.QuestionType;
 
-/** An immutable configuration of a program. */
+/**
+ * An immutable configuration of a program.
+ *
+ * <p>This class is serializable so that program definitions can be migrated between different
+ * instances of CiviForm, **not** so that it can be stored in the database. A program definition is
```
Comment (avaleske) [inline server/app/services/program/ProgramDefinition.java:40]: this is a good callout, thank you!

File: server/app/controllers/admin/ProgramMigration.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package controllers.admin;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import services.program.ProgramDefinition;
+
+/**
+ * A Jackson serializable class used to migrate programs between instances. {@link
```
Comment (avaleske) [inline server/app/controllers/admin/ProgramMigration.java:8]: **suggestion**: Can we say more about why this class is necessary? Is it a wrapper class to make serialization easier? If so can we name it like `ProgramMigrationWrapper`?

File: server/app/views/admin/migration/AdminImportViewPartial.java [hunk 4]
```diff
@@ -32,7 +40,36 @@ public DomContent renderError(String errorMessage) {
   }
 
   /** Renders the correctly parsed program data. */
-  public DomContent renderProgramData(String programData) {
-    return div(programData).withId(PROGRAM_DATA_ID);
+  public DomContent renderProgramData(ProgramMigration programMigration) {
+    ProgramDefinition program = programMigration.getProgram();
+    DivTag programDiv = div().withId(PROGRAM_DATA_ID);
+    programDiv.with(h3("Program name: " + program.localizedName().getDefault()));
+    programDiv.with(p("Admin name: " + program.adminName()));
```
Comment (avaleske) [inline server/app/views/admin/migration/AdminImportViewPartial.java:47]: optional nit: this is slightly cleaner (and similar reformatting is possible in `renderProgramBlock()` and `renderQuestion()`:

```suggestion
    DivTag programDiv = div().withId(PROGRAM_DATA_ID)
        .with(
            h3("Program name: " + program.localizedName().getDefault())
            p("Admin name: " + program.adminName()));
```

File: browser-test/src/admin/admin_program_migration.test.ts [hunk 5]
```diff
@@ -37,14 +81,18 @@ test.describe('program migration', {tag: ['@uses-fixtures']}, () => {
...
-      // TODO(#7087): We should also have a test that exports JSON and then imports that same JSON
-      // so that we can verify export and import work together.
       const sampleJson = readFileSync(
         'src/assets/import-program-sample.json',
         'utf8',
       )
       await adminProgramMigration.submitProgramJson(sampleJson)
 
+      await adminProgramMigration.expectProgramImported('Import Sample Program')
+      // The import page currently shows question IDs, so this screenshot needs
```
Comment (avaleske) [inline browser-test/src/admin/admin_program_migration.test.ts:91]: Does https://github.com/civiform/civiform/wiki/Testing#de-randomizing help with this?

File: server/test/services/program/ProgramDefinitionTest.java [hunk 6]
```diff
@@ -1247,4 +1253,180 @@ public void getSummaryImageFileKey_replacesExistingValue() {
     assertThat(def.summaryImageFileKey().get())
         .isEqualTo("program-summary-image/program-123/newFileKey.png");
   }
+
+  @Test
+  public void serializeThenDeserialize_allFieldsPresent() throws JsonProcessingException {
+    QuestionDefinition questionA = testQuestionBank.applicantName().getQuestionDefinition();
```
Comment (avaleske) [inline server/test/services/program/ProgramDefinitionTest.java:1259]: suggestion: should one of these be an enumerator question?

**Review by avaleske — APPROVED** [review]
(suggestion, nonblocking): It might be worth moving some of the serialization/deserialization logic into a `ProgramSerializationService`, so the controller can be lighter.

then you could use the same ObjectMapper in both as well

**Comment by caitlinshk** [comment]
Added `ProgramMigrationService` (trying to use "migration" everywhere) and addressed other PR comments.

File: server/app/controllers/admin/AdminExportController.java [hunk 7]
```diff
@@ -70,10 +77,30 @@ public Result exportProgram(Http.Request request) {
...
+    // TODO(#7087): The export UI only shows active programs. Should we not download the program
+    // JSON here if the programId is actually for a draft program?
+    ProgramDefinition program;
+    try {
+      program = programService.getFullProgramDefinition(programId);
+    } catch (ProgramNotFoundException e) {
+      return badRequest(String.format("Program with ID %s could not be found", programId));
+    }
+
+    ErrorAnd<String, String> serializeResult = programMigrationService.serialize(program);
```
Comment (avaleske) [inline server/app/controllers/admin/AdminExportController.java:96]: **_very_  nit and not a blocker, but to consider in future PRs**: the pattern elsewhere is for this to be a noun, like `serializationResult`

**Review by avaleske — APPROVED** [review]
thanks for the changes!

File: server/app/controllers/admin/AdminExportController.java (hunk 7 above)
Comment (avaleske) [reply server/app/controllers/admin/AdminExportController.java:96]: oh I just realized that's not a new class, just an `ErrorAnd`, nvm!

File: server/app/controllers/admin/ProgramMigration.java (hunk 1 above)
Comment (caitlinshk) [reply server/app/controllers/admin/ProgramMigration.java]: Oops yes good catch, updated!

**Pull request merged**